internal/pkg/cri/pods.go

```diff
@@ -104,21 +106,35 @@ func (c *Client) StopAndRemovePodSandboxes(ctx context.Context, stopAction StopA
 	for _, pod := range pods {
 		pod := pod // https://golang.org/doc/faq#closures_and_goroutines
 
-		status, _, err := c.PodSandboxStatus(ctx, pod.GetId())
-		if err != nil {
-			return err
-		}
+		g.Go(func() error {
+			status, _, e := c.PodSandboxStatus(ctx, pod.GetId())
+			if e != nil {
+				if grpcstatus.Code(e) == codes.NotFound {
+					return nil
+				}
 
-		networkMode := status.GetLinux().GetNamespaces().GetOptions().GetNetwork()
+				return e
+			}
 
-		// If any modes are specified, we verify that the current pod is
-		// running any one of the modes. If it doesn't, we skip it.
-		if len(modes) > 0 && !contains(networkMode, modes) {
-			continue
-		}
+			networkMode := status.GetLinux().GetNamespaces().GetOptions().GetNetwork()
 
-		g.Go(func() error {
-			return stopAndRemove(ctx, stopAction, c, pod, networkMode.String())
+			// If any modes are specified, we verify that the current pod is
+			// running any one of the modes. If it doesn't, we skip it.
+			if len(modes) > 0 && !contains(networkMode, modes) {
+				return nil
+			}
+
+			if pod.GetState() != runtimeapi.PodSandboxState_SANDBOX_READY {
+				log.Printf("skipping pod %s/%s, state %s", pod.Metadata.Namespace, pod.Metadata.Name, pod.GetState())
```

Comment:
This pod will ultimately get cleaned up? My only concern is that we need to be certain that all pods are removed in case they mount something that prevents our upgrade.

Reply:
I believe if sandbox is not ready, sandbox is stopped. 

So the flow with single-node upgrade is the following:

* first we stop all the pods/containers which depend on the CNI
* then we try stopping all the pods (including NODE networking), but that also goes once again over the pods we stopped in the previous step, so this check prevents a pod being stopped twice, which with canal results in CNI error, as it tries to reach out to the CNI which is already stopped

As for the mounts, we check for them explicitly before trying an upgrade.

Reply:
Cool. Yeah I know we check, but I want to be sure we can handle them gracefully.

Reply:
yep, if it's not a single-node case, pods are removed, so by the time we get to this step, pods is not in the list even